doc/languages-frameworks/rust.section.md

```diff
@@ -261,6 +264,43 @@ general. A number of other parameters can be overridden:
   };
   ```
 
+### Setting Up `nix-shell`
+Oftentimes you want to develop code using `nix-shell`. Using the example above,
+we want to do the following:
+
+- Have access to `cargo` and `rustc`
+- Have the `openssl` library available to the rust crate through
+  it's _normal_ compilation mechanism (`pkg-config`).
+
+`default.nix`:
+```
+with import <nixpkgs> {};
+((import ./hello.nix).hello {}).override {
+  crateOverrides = defaultCrateOverrides // {
+    hello = attrs: {
+        buildInputs = [
+        # Some example external dependencies
+        openssl xorg.libX11 xorg.libXtst
+        # For `nix-shell`
+        rustc cargo openssl pkgconfig
+      ];
+
+      # Set Relevant ENV Variables
+      RUST_BACKTRACE=1;
+    };
+  };
+}
+```
+
+You should now be able to run the following:
+```
+$ nix-shell --pure
+$ cargo build
+$ cargo test
+```
+
+Crates that have the above dependencies will still compile.
+
```

Comment:
That's indeed a very good point to add to the documentation.

However, this is not what I usually do. Your code here would compile crate `hello` and put its binaries in the scope of the shell. I wonder why you would do that.

If it helps, my usual default.nix uses the builds from rustup (mostly because they're more current), even though this doesn't work on MacOS:

```nix
with import <nixpkgs> {};

let src = fetchFromGitHub {
      owner = "mozilla";
      repo = "nixpkgs-mozilla";
      rev = "42a0926f2f36cac2da53782259948ba071b6c6c5";
      sha256 = "1r2jglgl9k881byv1kc3rdda2lzaarvb0xn7nx3q0b3h25apjff5";
   };
in
with import "${src.out}/rust-overlay.nix" pkgs pkgs;

stdenv.mkDerivation {
  name = "rust-env";
  buildInputs = [ rustChannels.stable.rust ];
}
```

Reply:
Another option (MacOS-compatible, but you won't get the nightly builds) is simply:

```
with import <nixpkgs> {};

stdenv.mkDerivation {
  name = "rust-env";
  buildInputs = [ rust ];
}
```

Reply:
So I think there is some confusion here...

I wanted a `default.nix` which can build the package and _simultaneously_ allows me to use `nix-shell`. I don't think what you posted solves that problem, right?

Reply:
Right, sorry about that. I usually have a default.nix and a shell.nix, but what you wrote is another solution.
Have you tested though, that `RUST_BACKTRACE` is actually set in the resulting shell? I don't remember that any variable besides the standard ones is inherited in this way, feel free to update build-rust-crate.nix if not.

Reply:
Hmm... it doesn't seem to work. I _swear_ I tested it -- but I was also simultaneously testing `configuration.nix`. I probably set it there and didn't use `--pure` and got confused :(

I think this all goes back to the core point: I should use `stdenv.mkDerivation` instead for shell right? I will get right on that :)

Reply:
one thing that is annoying though: it would be _great_ to have the build and test/shell environment be identical. Do you know of a way for `nix-build` to run rust tests (or tests in general)?

Reply:
Yes, there is a test phase in standard derivations, but I've not implemented that yet in `build-rust-crate.nix`. I suppose I could work on that some day, but not right now. Meanwhile, you could try to add something in `build-rust-crate.nix` to inherit all extra variables one might set in an overrideCrate, such as `RUST_BACKTRACE`.

Reply:
Okay, I think I will just rework the docs noting that those things are not yet supported for `nix-shell` and how to use `stdenv.mkDerivation` instead.

`nix-shell` still seems magical to me -- how does it know what things are "setup time" vs "build time" (rhetorical)? Other newbies will probably have similar questions, so I think it is a good place to admit our faults :)

Reply:
Ok, I didn't remember this was an issue, so PLEASE add it to the current documentation: a derivation has phases, documented in https://nixos.org/nixpkgs/manual/#sec-stdenv-phases.

Nix-build sets up the environment by downloading all the buildInputs/propagatedBuildInputs, and then runs all the phases, while nix-shell only sets up the environment.

Reply:
sounds good. I will reference the issue I just opened: #37945

Reply:
Note: your example of `buildInputs = [ rust ]` gives me this:

```
error: cannot coerce a set to a string, at /nix/store/y68vv1sxb24yykccac566v4xqklnzfa2-nixos-18.03pre130932.cc4677c36ee/nixos/pkgs/stdenv/generic/make-derivation.nix:148:11
(use '--show-trace' to show detailed location information)
```

changing to `rustc cargo` works.